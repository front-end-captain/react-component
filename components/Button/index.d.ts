import * as React from "react";

export declare type ButtonType = "default" | "primary" | "dashed" | "danger";
export declare type ButtonSize = "large" | "default" | "small";
export declare type ButtonHtmlType = "button" | "submit" | "reset";

export interface ButtonProps {
  disabled?: boolean;
  loading?: boolean;
  type?: ButtonType;
  size?: ButtonSize;
  prefixCls?: string;
  className?: string;
  htmlType?: ButtonHtmlType;
  onClick?: React.MouseEventHandler<HTMLAnchorElement>;
  children?: React.ReactNode;
}

declare const Button: (props: ButtonProps) => React.ReactNode;

export default Button;
