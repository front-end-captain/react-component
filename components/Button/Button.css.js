import styled from "styled-components";
import { primaryColor, dangerColor, defaultBorderColor, fontSize } from "./../_style/vars.js";

const ButtonWrapper = styled.button`
  line-height: 1.499;
  display: inline-block;
  font-weight: 400;
  text-align: center;
  touch-action: manipulation;
  cursor: pointer;
  outline: none;
  background-image: none;
  border: 1px solid transparent;
  white-space: nowrap;
  padding: 0 15px;
  font-size: ${fontSize};
  border-radius: 4px;
  height: 32px;
  user-select: none;
  transition: all 0.3s cubic-bezier(0.645, 0.045, 0.355, 1);
  position: relative;
  box-shadow: 0 2px 0 rgba(0, 0, 0, 0.015);
  color: rgba(0, 0, 0, 0.65);
  background-color: #fff;
  border-color: ${defaultBorderColor};

  &.${(props) => props.prefixCls}-danger {
    color: ${dangerColor};
    background-color: #f5f5f5;
    border-color: ${defaultBorderColor};
  }

  &.${(props) => props.prefixCls}-danger:hover {
    color: #fff;
    background-color: #ff4d4f;
    border-color: #ff4d4f;
  }

  &.${(props) => props.prefixCls}-primary {
    color: #fff;
    background-color: ${primaryColor};
    border-color: #1890ff;
    text-shadow: 0 -1px 0 rgba(0, 0, 0, 0.12);
    box-shadow: 0 2px 0 rgba(0, 0, 0, 0.045);
  }

  &.${(props) => props.prefixCls}-primary:hover {
    color: #fff;
    background-color: #40a9ff;
    border-color: #40a9ff;
  }

  &.${(props) => props.prefixCls}-dashed {
    color: rgba(0, 0, 0, 0.65);
    background-color: #fff;
    border-color: ${defaultBorderColor};
    border-style: dashed;
  }

  &.${(props) => props.prefixCls}-dashed:hover {
    color: #40a9ff;
    background-color: #fff;
    border-color: #40a9ff;
  }

  &.${(props) => props.prefixCls}[disabled] {
    cursor: not-allowed;
    color: rgba(0, 0, 0, 0.25);
    background-color: #f5f5f5;
    border-color: ${defaultBorderColor};
    text-shadow: none;
    box-shadow: none;
  }
`;

export default ButtonWrapper;
