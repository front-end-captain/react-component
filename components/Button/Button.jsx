import React from "react";
import PropTypes from "prop-types";
import classNames from "classnames";

import { LoadingIcon } from "./../Icon/index.jsx";

import ButtonWrapper from "./Button.css.js";

const sizes = { small: "small", default: "default", large: "large" };

const types = { primary: "primary", default: "default", danger: "danger", dashed: "dashed" };

const htmlTypes = { button: "button", submit: "submit", reset: "reset" };

const Button = ({
  children,
  className,
  onClick,
  disabled,
  loading,
  htmlType,
  type,
  size,
  prefixCls,
  block,
  ...resetProps
}) => {
  const isDisabled = disabled || loading ? { disabled: true } : { onClick };
  const baseProps = {
    ...resetProps,
    ...isDisabled,
    prefixCls,
    type: htmlType,
    className: classNames(prefixCls, className, {
      [`${prefixCls}-${type}`]: type,
      [`${prefixCls}-default`]: !disabled && type === types.default,
      [`${prefixCls}-normal`]: type === types.default,
      [`${prefixCls}-disabled`]: disabled,
      [`${prefixCls}-loading`]: loading,
      [`${prefixCls}-size-${size}`]: size !== sizes.default,
      [`${prefixCls}-block`]: block,
    }),
  };
  return (
    <ButtonWrapper {...baseProps}>
      {loading && <LoadingIcon className="sky-loading" />}
      <span>{children || "Button"}</span>
    </ButtonWrapper>
  );
};

Button.defaultProps = {
  prefixCls: "sky-btn",
  loading: false,
  type: types.default,
  size: sizes.default,
  htmlType: htmlTypes.button,
  disabled: false,
  block: false,
};

export const ButtonPropTypes = {
  prefixCls: PropTypes.string,
  loading: PropTypes.bool,
  type: PropTypes.oneOf(Object.values(types)),
  size: PropTypes.oneOf(Object.values(sizes)),
  htmlType: PropTypes.oneOf(Object.values(htmlTypes)),
  disabled: PropTypes.bool,
  block: PropTypes.bool,
};

Button.propTypes = ButtonPropTypes;

export default Button;
