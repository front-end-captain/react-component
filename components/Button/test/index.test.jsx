import React from "react";
import assert from "power-assert";
import { render, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import Button from "../index";
import { CircleLoadingIcon } from "./../../Icon/index.jsx";

describe("<Button/>", () => {
  it("should render a <Button/> components", () => {
    const wrapper = render(
      <div>
        <Button>Hello Sky</Button>
        <Button type="primary">Primary</Button>
        <Button type="dashed">Dashed</Button>
        <Button type="danger">Danger</Button>
        <Button disabled>Disabled</Button>
        <Button loading>loading</Button>
        <Button block>Block</Button>
        <Button type="primary" size="large">
          Large
        </Button>
        <Button type="primary">Default</Button>
        <Button type="danger" size="small">
          Small
        </Button>
      </div>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should find sky-button classnames", () => {
    const wrapper = render(
      <div>
        <Button>你好</Button>
        <Button type="primary">Primary</Button>
        <Button type="default">Default</Button>
        <Button type="danger">Danger</Button>
        <Button type="dashed">Dashed</Button>
        <Button disabled>Disabled</Button>
        <Button loading>loading</Button>
        <Button block>Block</Button>
        <Button type="primary" size="large">
          Primary
        </Button>
      </div>,
    );
    assert(wrapper.find(".sky-btn").length >= 1);
    assert(wrapper.find(".sky-btn-primary").length === 2);
    assert(wrapper.find(".sky-btn-danger").length === 1);
    assert(wrapper.find(".sky-btn-default").length === 4);
    assert(wrapper.find(".sky-btn-disabled").length === 1);
    assert(wrapper.find(".sky-btn-loading").length === 1);
    assert(wrapper.find(".sky-btn-block").length === 1);
    assert(wrapper.find(".sky-btn-dashed").length === 1);
  });

  it("should can trigger click event", () => {
    const onClick = jest.fn();
    const wrapper = shallow(<Button onClick={onClick}>sky</Button>);
    wrapper.find(".sky-btn").simulate("click");
    expect(onClick).toHaveBeenCalled();
  });

  it("should can not trigger click event", () => {
    const onClick = jest.fn();
    const wrapper = shallow(
      <Button onClick={onClick} disabled>
        sky
      </Button>,
    );
    wrapper.find(".sky-btn").simulate("click");
    expect(onClick).not.toHaveBeenCalled();
  });

  it("should can render loading icon with button", () => {
    const wrapper = shallow(<Button loading>sky</Button>);
    assert(wrapper.find(CircleLoadingIcon).length === 1);
    assert(wrapper.find(".sky-loading").length === 1);
  });
});
