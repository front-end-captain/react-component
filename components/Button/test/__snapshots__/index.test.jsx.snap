// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button/> should render a <Button/> components 1`] = `
<div>
  <button
    class="sky-btn sky-btn-default sky-btn-normal Buttoncss__ButtonWrapper-k9s1tq-0 cRNwTE"
    type="button"
  >
    <span>
      Hello Sky
    </span>
  </button>
  <button
    class="sky-btn sky-btn-primary Buttoncss__ButtonWrapper-k9s1tq-0 cRNwTE"
    type="button"
  >
    <span>
      Primary
    </span>
  </button>
  <button
    class="sky-btn sky-btn-dashed Buttoncss__ButtonWrapper-k9s1tq-0 cRNwTE"
    type="button"
  >
    <span>
      Dashed
    </span>
  </button>
  <button
    class="sky-btn sky-btn-danger Buttoncss__ButtonWrapper-k9s1tq-0 cRNwTE"
    type="button"
  >
    <span>
      Danger
    </span>
  </button>
  <button
    class="sky-btn sky-btn-normal sky-btn-disabled Buttoncss__ButtonWrapper-k9s1tq-0 cRNwTE"
    disabled=""
    type="button"
  >
    <span>
      Disabled
    </span>
  </button>
  <button
    class="sky-btn sky-btn-default sky-btn-normal sky-btn-loading Buttoncss__ButtonWrapper-k9s1tq-0 cRNwTE"
    disabled=""
    type="button"
  >
    <svg
      class="sky-loading"
      fill="currentColor"
      height="1em"
      version="1.1"
      viewBox="0 0 1024 1024"
      width="1em"
      xlink="http://www.w3.org/1999/xlink"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs />
      <path
        d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 0 0-94.3-139.9 437.71 437.71 0 0 0-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3 0.1 19.9-16 36-35.9 36z"
        p-id="1345"
      />
    </svg>
    <span>
      loading
    </span>
  </button>
  <button
    class="sky-btn sky-btn-default sky-btn-normal sky-btn-block Buttoncss__ButtonWrapper-k9s1tq-0 cRNwTE"
    type="button"
  >
    <span>
      Block
    </span>
  </button>
  <button
    class="sky-btn sky-btn-primary sky-btn-size-large Buttoncss__ButtonWrapper-k9s1tq-0 cRNwTE"
    type="button"
  >
    <span>
      Large
    </span>
  </button>
  <button
    class="sky-btn sky-btn-primary Buttoncss__ButtonWrapper-k9s1tq-0 cRNwTE"
    type="button"
  >
    <span>
      Default
    </span>
  </button>
  <button
    class="sky-btn sky-btn-danger sky-btn-size-small Buttoncss__ButtonWrapper-k9s1tq-0 cRNwTE"
    type="button"
  >
    <span>
      Small
    </span>
  </button>
</div>
`;
